spread/image/create-image.sh

```diff
@@ -0,0 +1,147 @@
+ubuntu_image_extra_args=
+
+if [ ! -z "$snap" ] ; then
+	ubuntu_image_extra_args="--extra-snaps $snap"
+fi
+
+ubuntu-image \
+	--channel $channel \
+	-o $image_name \
+	--image-size 4G \
+	$ubuntu_image_extra_args \
+	$model.model
+
+kpartx -a $image_name
+sleep 0.5
+
+loop_path=`findfs LABEL=writable`
+tmp_mount=`mktemp -d`
+
+mount $loop_path $tmp_mount
+
+# Migrate all systemd units from core snap into the writable area. This
+# would be normally done on firstboot by the initramfs but we can't rely
+# on that because we  are adding another file in there and that will
+# prevent the initramfs from transitioning any files.
+core_snap=$(find $tmp_mount/system-data/var/lib/snapd/snaps -name "core_*.snap")
+tmp_core=`mktemp -d`
+mount $core_snap $tmp_core
+mkdir -p $tmp_mount/system-data/etc/systemd
+cp -rav $tmp_core/etc/systemd/* \
+	$tmp_mount/system-data/etc/systemd/
+umount $tmp_core
+rm -rf $tmp_core
+
+# system-user assertion which gives us our test:test user we use to
+# log into the system
+mkdir -p $tmp_mount/system-data/var/lib/snapd/seed/assertions
+cp test-user.assertion $tmp_mount/system-data/var/lib/snapd/seed/assertions
+
+# Disable console-conf for the first boot
+mkdir -p $tmp_mount/system-data/var/lib/console-conf/
+touch $tmp_mount/system-data/var/lib/console-conf/complete
+
+# Create systemd service which is running on firstboot and sets up
+# various things for us.
+mkdir -p $tmp_mount/system-data/etc/systemd/system
+cat << 'EOF' > $tmp_mount/system-data/etc/systemd/system/devmode-firstboot.service
+[Unit]
+Description=Run devmode firstboot setup
+After=snapd.service snapd.socket
+
+[Service]
+Type=oneshot
+ExecStart=/writable/system-data/var/lib/devmode-firstboot/run.sh
+RemainAfterExit=yes
+TimeoutSec=3min
+EOF
+
+mkdir -p $tmp_mount/system-data/etc/systemd/system/multi-user.target.wants
+ln -sf /etc/systemd/system/devmode-firstboot.service \
+	$tmp_mount/system-data/etc/systemd/system/multi-user.target.wants/devmode-firstboot.service
+
+mkdir $tmp_mount/system-data/var/lib/devmode-firstboot
+cat << 'EOF' > $tmp_mount/system-data/var/lib/devmode-firstboot/run.sh
+#!/bin/bash
+
+set -e
+
+# Don't start again if we're already done
+if [ -e /writable/system-data/var/lib/devmode-firstboot/complete ] ; then
+	exit 0
+fi
+
+echo "Start devmode-firstboot $(date -Iseconds --utc)"
+
+if [ "$(snap managed)" = "true" ]; then
+	echo "System already managed, exiting"
+	exit 0
+fi
+
+# no changes at all
+while ! snap changes ; do
+	echo "No changes yet, waiting"
+	sleep 1
+done
+
+while snap changes | grep -qE '(Do|Doing) .*Initialize system state' ;  do
+	echo "Initialize system state is in progress, waiting"
+	sleep 1
+done
+
+if [ -n "$(snap known system-user)" ]; then
+	echo "Trying to create known user"
+	snap create-user --known --sudoer
+fi
+
+# Enable console-conf again
+rm /writable/system-data/var/lib/console-conf/complete
+
+# Mark us done
+touch /writable/system-data/var/lib/devmode-firstboot/complete
+
```

Comment:
We can probably add below lines to ensure our ci ssh keys are in the authorized list on first run.

echo ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDQhliDv2ql/5YW/XE6nKd8Do9lhzeLC6ubUD8qO294FQ3hJdxhXJhJnSTmDbvDtzyviYhRitw03+pnKLv7SMBlvReX1QwsaxJdflW7pP8tll5tbqDCiLv66KewApLUQuPXAtgJCtYlwyOhHQNbw1fs37AsicmcKgpr0ow2GpNqgwCNwf8cizCD4XceafeJ4pwO+LTM1Y4uXqsFPfY21t1PVxKDwhcZ/4P/9zheyrMP+5xNsFBzUWMM9icKJUn3YDXRjaTIOB8+eam5xuqJZC1Ig4ZjBrRcp3vohDvsVccTcLXzbdBFNIxao/ZhrMW/PeKLcvVY+z4gX3Ev3eqx4GOP snapweb.ci@gmail.com > /home/test/.ssh/authorized_keys